h2o-persist-hdfs/src/main/java/water/persist/PersistHdfs.java

```diff
@@ -286,24 +282,39 @@ private static void ignoreAndWait(final Exception e, boolean printException) {
   }
 
   public static void addFolder(Path p, ArrayList<String> keys,ArrayList<String> failed) throws IOException {
-    FileSystem fs = FileSystem.get(p.toUri(), PersistHdfs.CONF);
+    FileSystem fs = null;
+    final AmazonCredentials amazonCredentials = extractAmazonCredentials(p.toString());
+    try {
+      fs = FileSystem.get(new URI(p.toString()), PersistHdfs.CONF);
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException("Given S3 URI is malformed. Please check special characters and encoding.", e);
```

Comment:
This class deals with all Hadoop filesystems, not just S3. Please also include the path/(=user input) in the message.